README.md
# Welcome to Ideas Portal !

Extension code for Ingram Micro Ideas (AHA)

## Service configuration
### Enviroment variables

* AHA_JWT_SECRET: Needs to be set to be able to encode jwt.
* AHA_LOGIN_URL: URL for aha login.
* APPROVED_TEMPLATE_ID: Id of the approved template that we need to use on purchase process.
* TOKEN_EXP_MINUTES: By default token's expiration is set in one minute. But we can change it by setting this variable. Units = minutes.
* DEFAULT_REDIRECT: By default is set to ingrammicrocloud.com
## License

**Ideas Portal** is licensed under the *Apache Software License 2.0* license.



CHANGELOG.md
# EaaS: Ideas Portal changelog

* 1.0.0: initial version


tests/conftest.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021, Ingram Micro Cloud
# All rights reserved.
#

import time
from collections import namedtuple
from collections.abc import Iterable
from types import MethodType
from urllib.parse import parse_qs

from connect.client import ConnectClient

from connect_ext.extension import IdeasPortalExtension

from jwt import encode

import pytest

import requests

import responses


ConnectResponse = namedtuple(
    'ConnectResponse',
    (
        'count', 'query', 'ordering', 'select',
        'value', 'status', 'exception',
    ),
)


def _parse_qs(url):
    if '?' not in url:
        return None, None, None

    url, qs = url.split('?')
    parsed = parse_qs(qs, keep_blank_values=True)
    ordering = None
    select = None
    query = None

    for k in parsed.keys():
        if k.startswith('ordering('):
            ordering = k[9:-1].split(',')
        elif k.startswith('select('):
            select = k[7:-1].split(',')
        else:
            value = parsed[k]
            if not value[0]:
                query = k

    return query, ordering, select


def _mock_kwargs_generator(response_iterator, url):
    res = next(response_iterator)

    query, ordering, select = _parse_qs(url)
    if res.query:
        assert query == res.query, 'RQL query does not match.'
    if res.ordering:
        assert ordering == res.ordering, 'RQL ordering does not match.'
    if res.select:
        assert select == res.select, 'RQL select does not match.'
    mock_kwargs = {
        'match_querystring': False,
    }
    if res.count is not None:
        end = 0 if res.count == 0 else res.count - 1
        mock_kwargs['status'] = 200
        mock_kwargs['headers'] = {'Content-Range': f'items 0-{end}/{res.count}'}
        mock_kwargs['json'] = []

    mock_kwargs.update(_value_arg_validation(res))
    return mock_kwargs


def _value_arg_validation(res):
    result = {}
    if isinstance(res.value, Iterable):
        count = len(res.value)
        end = 0 if count == 0 else count - 1
        result['status'] = 200
        result['json'] = res.value
        result['headers'] = {
            'Content-Range': f'items 0-{end}/{count}',
        }
    elif isinstance(res.value, dict):
        result['status'] = res.status or 200
        result['json'] = res.value
    elif res.value is None:
        if res.exception:
            result['body'] = res.exception
        else:
            result['status'] = res.status
    else:
        result['status'] = res.status or 200
        result['body'] = str(res.value)
    return result


@pytest.fixture
def response():
    with responses.RequestsMock() as rsps:
        yield rsps


@pytest.fixture
def logger(mocker):
    return mocker.MagicMock()


@pytest.fixture
def response_factory():
    def _create_response(
        count=None,
        query=None,
        ordering=None,
        select=None,
        value=None,
        status=None,
        exception=None,
    ):
        return ConnectResponse(
            count=count,
            query=query,
            ordering=ordering,
            select=select,
            value=value,
            status=status,
            exception=exception,
        )
    return _create_response


@pytest.fixture
def sync_client_factory():
    def _create_sync_client(connect_responses):
        response_iterator = iter(connect_responses)

        def _execute_http_call(self, method, url, kwargs):
            mock_kwargs = _mock_kwargs_generator(response_iterator, url)
            with responses.RequestsMock() as rsps:
                rsps.add(
                    method.upper(),
                    url,
                    **mock_kwargs,
                )
                self.response = requests.request(method, url, **kwargs)
                if self.response.status_code >= 400:
                    self.response.raise_for_status()

        client = ConnectClient('Key', use_specs=False)
        client._execute_http_call = MethodType(_execute_http_call, client)
        return client
    return _create_sync_client


@pytest.fixture
def product_action_request_factory():
    def _product_action_request_factory(method, form_data=None):
        form_data = form_data if form_data else {}
        jwt_payload = {
            'exp': time.time(),
            'asset_id': 'AS-1111-2222-3333',
            'action_id': 'sso_action',
        }

        connect_token = encode(jwt_payload, "THE_SECRET")
        return {
            'method': method,
            'querystring': {'jwt': [connect_token]},
            'form_data': {
                'email': form_data.get('email', 'john.doe@example.com'),
                'givenName': form_data.get('givenName', 'John'),
                'familyName': form_data.get('familyName', 'Doe'),
            },
            'jwt_payload': jwt_payload,
        }
    return _product_action_request_factory


@pytest.fixture
def product_action_config_factory():
    def _product_action_request_factory(expiration_time_conf=1):
        return {
            'AHA_LOGIN_URL': 'https://imc.ideas.aha.io/auth/jwt/callback/',
            'AHA_JWT_SECRET': "AHA_SECRET_KEY",
            'TOKEN_EXP_MINUTES': expiration_time_conf,
            'DEFAULT_REDIRECT': 'https://ingrammicrocloud.com',
            'APPROVED_TEMPLATE_ID': "approved-temp-id",
        }
    return _product_action_request_factory


@pytest.fixture
def extension_for_asset_approval_factory(sync_client_factory, response_factory, logger):
    def _extension_for_asset_aproval_factory(http_status_code=200, exception=None):
        config = {'APPROVED_TEMPLATE_ID': 'DUMMY-TEMPLATE-ID'}
        responses = [
            response_factory(status=http_status_code, exception=exception),
        ]
        client = sync_client_factory(responses)
        return IdeasPortalExtension(client, logger, config)

    return _extension_for_asset_aproval_factory


tests/test_ideasportal.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021, Globex Corporation
# All rights reserved.
#

import re
import time

import connect

from connect_ext.extension import IdeasPortalExtension

import pytest

from jwt import decode

from freezegun import freeze_time

TEST_FREEZE_TIME_EPOCH = 1626873720
TEST_FREEZE_TIME = '2021-07-21 13:22:00'


def test_process_asset_purchase_request_raise_error(
    sync_client_factory,
    response_factory,
    logger,
):
    config = {'APPROVED_TEMPLATE_ID': 'DUMMY-TEMPLATE-ID'}
    request = {'id': 1, 'status': 'pending', 'type': 'purchase'}
    responses = [
        response_factory(status=400),
    ]
    client = sync_client_factory(responses)
    ext = IdeasPortalExtension(client, logger, config)

    with pytest.raises(connect.client.exceptions.ClientError) as excinfo:
        ext.process_asset_purchase_request(request)
        assert '400 Bad Request' in excinfo.value


def test_process_asset_purchase_request_raise_500_error(
    sync_client_factory,
    response_factory,
    logger,
):
    config = {}
    request = {'id': 1, 'status': 'pending', 'type': 'purchase'}
    responses = [
        response_factory(status=400),
    ]
    client = sync_client_factory(responses)
    ext = IdeasPortalExtension(client, logger, config)

    with pytest.raises(Exception) as excinfo:
        ext.process_asset_purchase_request(request)
        assert '500 Server Error' in excinfo.value


@pytest.mark.parametrize(
    'request_method,redirect_url_key,query_string_pattern',
    [('POST', 'AHA_LOGIN_URL', '\\?jwt=*'),
     ('GET', 'DEFAULT_REDIRECT', '')],
)
def test_process_product_action(
    logger,
    product_action_request_factory,
    product_action_config_factory,
    request_method,
    redirect_url_key,
    query_string_pattern,
):
    config = product_action_config_factory()

    request = product_action_request_factory(request_method)
    ext = IdeasPortalExtension(None, logger, config)
    result = ext.execute_product_action(request)
    redirect_url = config[redirect_url_key]

    assert result.status == 'success'
    assert result.http_status == 302
    assert re.match(re.compile(f'{redirect_url}{query_string_pattern}'), result.headers['Location'])


@pytest.mark.parametrize(
    'expiration_time_conf,expected_expiration',
    [(5, TEST_FREEZE_TIME_EPOCH + 5 * 60),
     (0, TEST_FREEZE_TIME_EPOCH),
     ]
)
@freeze_time(TEST_FREEZE_TIME)
def test_process_product_action_aha_payload(
    logger,
    product_action_request_factory,
    product_action_config_factory,
    expiration_time_conf,
    expected_expiration,
):
    config = product_action_config_factory()
    config['TOKEN_EXP_MINUTES'] = expiration_time_conf

    request = product_action_request_factory('POST')

    ext = IdeasPortalExtension(None, logger, config)
    result = ext.execute_product_action(request)
    redirect_url = config['AHA_LOGIN_URL']
    location = result.headers['Location']

    assert result.status == 'success'
    assert result.http_status == 302

    url_pattern = re.compile(f'{redirect_url}\?jwt=(.+)')
    assert re.match(url_pattern, location)

    freeze_moment = time.time()
    jwt_token = url_pattern.search(location).group(1)
    jwt_paylod_result = decode(jwt_token, config['AHA_JWT_SECRET'], algorithms='HS256')
    request_form_data = request['form_data']
    assert freeze_moment == jwt_paylod_result['iat']
    assert request['jwt_payload']['asset_id'] == jwt_paylod_result['jti']
    assert request_form_data['givenName'] == jwt_paylod_result['first_name']
    assert request_form_data['familyName'] == jwt_paylod_result['last_name']
    assert request_form_data['email'] == jwt_paylod_result['email']
    assert expected_expiration == jwt_paylod_result['exp']


tests/test_product_action.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021, Ingram Micro Cloud
# All rights reserved.
#

import re
import time

from connect_ext.extension import IdeasPortalExtension

import pytest

from jwt import decode

from freezegun import freeze_time

TEST_FREEZE_TIME_EPOCH = 1626873720
TEST_FREEZE_TIME = '2021-07-21 13:22:00'


@pytest.mark.parametrize(
    ('request_method', 'redirect_url_key', 'query_string_pattern'),
    (
        ('POST', 'AHA_LOGIN_URL', '\\?jwt=*'),
        ('GET', 'DEFAULT_REDIRECT', ''),
    ),
)
def test_process_product_action(
    logger,
    product_action_request_factory,
    product_action_config_factory,
    request_method,
    redirect_url_key,
    query_string_pattern,
):
    config = product_action_config_factory()

    request = product_action_request_factory(request_method)
    ext = IdeasPortalExtension(None, logger, config)
    result = ext.execute_product_action(request)
    redirect_url = config[redirect_url_key]

    assert result.status == 'success'
    assert result.http_status == 302
    assert re.match(re.compile(f'{redirect_url}{query_string_pattern}'), result.headers['Location'])


@pytest.mark.parametrize(
    ('expiration_time_conf', 'expected_expiration'),
    (
        (5, TEST_FREEZE_TIME_EPOCH + 5 * 60),
        (0, TEST_FREEZE_TIME_EPOCH),
    ),
)
@freeze_time(TEST_FREEZE_TIME)
def test_process_product_action_aha_payload(
    logger,
    product_action_request_factory,
    product_action_config_factory,
    expiration_time_conf,
    expected_expiration,
):
    config = product_action_config_factory(expiration_time_conf)
    request = product_action_request_factory('POST')

    ext = IdeasPortalExtension(None, logger, config)
    result = ext.execute_product_action(request)
    redirect_url = config['AHA_LOGIN_URL']
    location = result.headers['Location']

    assert result.status == 'success'
    assert result.http_status == 302

    url_pattern = re.compile(f'{redirect_url}\\?jwt=(.*)')
    assert re.match(url_pattern, location)

    freeze_moment = time.time()
    jwt_token = url_pattern.search(location).group(1)
    jwt_paylod_result = decode(jwt_token, config['AHA_JWT_SECRET'], algorithms='HS256')
    request_form_data = request['form_data']
    assert freeze_moment == jwt_paylod_result['iat']
    assert request['jwt_payload']['asset_id'] == jwt_paylod_result['jti']
    assert request_form_data['givenName'] == jwt_paylod_result['first_name']
    assert request_form_data['familyName'] == jwt_paylod_result['last_name']
    assert request_form_data['email'] == jwt_paylod_result['email']
    assert expected_expiration == jwt_paylod_result['exp']


@pytest.mark.parametrize(
    ('form_data', 'redirect_url_key', 'query_string_pattern'),
    (
        (
            {'email': '',
             'givenName': 'John',
             'familyName': 'Doe',
             },
            'DEFAULT_REDIRECT',
            '',
        ),
        (
            {'email': 'john.doe@cloudblue.com',
             'givenName': '',
             'familyName': '',
             },
            'AHA_LOGIN_URL',
            '\\?jwt=*',
        ),
    ),
)
def test_process_product_action_aha_form_data(
    logger,
    product_action_request_factory,
    product_action_config_factory,
    form_data,
    redirect_url_key,
    query_string_pattern,
):
    config = product_action_config_factory()

    request = product_action_request_factory(method='POST', form_data=form_data)

    ext = IdeasPortalExtension(None, logger, config)
    result = ext.execute_product_action(request)
    redirect_url = config[redirect_url_key]

    assert result.status == 'success'
    assert result.http_status == 302
    assert re.match(re.compile(f'{redirect_url}{query_string_pattern}'), result.headers['Location'])


connect_ext/extension.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021, Ingram Micro Cloud
# All rights reserved.
#

import time

from connect.client.exceptions import ClientError
from connect.eaas.extension import (
    Extension,
    ProcessingResponse,
    ProductActionResponse,
)

from jwt import encode


class IdeasPortalExtension(Extension):

    DEFAULT_REDIRECT = 'https://ingrammicrocloud.com'
    DEFAULT_EXPIRATION_MINUTES = 1

    def _calculate_aha_token(self, data, jwt_payload):
        expiration_mins = self.config.get("TOKEN_EXP_MINUTES", self.DEFAULT_EXPIRATION_MINUTES)

        payload = {}
        payload['iat'] = int(time.time())
        payload['jti'] = jwt_payload['asset_id']
        payload['first_name'] = data.get('givenName', '')
        payload['last_name'] = data.get('familyName', '')
        payload['email'] = data['email']
        payload['exp'] = time.time() + expiration_mins * 60

        return encode(payload, self.config['AHA_JWT_SECRET'])

    def _execute_asset_approval(self, request):
        request_id = request['id']
        request_type = request['type']
        template_id = self.config['APPROVED_TEMPLATE_ID']
        customer_id = request['asset']['tiers']['customer']['external_id']
        asset_id = request['asset']['id']

        self.logger.info(
            f'Processing request {request_id} of type {request_type} for asset {asset_id} '
            f'related with reseller {customer_id}',
        )

        try:
            self.client.requests[request_id]('approve').post({'template_id': template_id})
            self.logger.info(
                f'Successfully processed request {request_id} of type {request_type} for asset '
                f'{asset_id} related with reseller {customer_id}',
            )
            return ProcessingResponse.done()
        except ClientError as exception:
            self.logger.error(
                f'Error processing request {request_id} of type {request_type} for asset '
                f'{asset_id} related with reseller {customer_id}',
            )
            if exception.status_code and exception.status_code >= 500:
                return ProcessingResponse.reschedule()
            raise exception

    def process_asset_purchase_request(self, request):
        return self._execute_asset_approval(request)

    def process_asset_cancel_request(self, request):
        return self._execute_asset_approval(request)

    def process_asset_change_request(self, request):
        return self._execute_asset_approval(request)

    def process_asset_resume_request(self, request):
        return self._execute_asset_approval(request)

    def process_asset_suspend_request(self, request):
        return self._execute_asset_approval(request)

    def execute_product_action(self, request):
        jwt_payload = request['jwt_payload']
        asset_id = jwt_payload['asset_id']
        location = self.config.get('DEFAULT_REDIRECT', self.DEFAULT_REDIRECT)

        if request['method'] == 'POST':
            data = request.get('form_data', {})
            email = data.get('email')
            if not email:
                self.logger.error(f'SSO requested for asset {asset_id} with no email')
                self.logger.info(
                    f'Redirecting SSO request for asset {asset_id} to default site',
                )
            else:
                self.logger.info(f'SSO requested for asset {asset_id} by {email}')
                aha_token = self._calculate_aha_token(data, jwt_payload)
                aha_login_url = self.config['AHA_LOGIN_URL']
                location = f'{aha_login_url}?jwt={aha_token}'
                self.logger.info(
                    f'Redirecting SSO request for asset {asset_id} to login',
                )
        else:
            self.logger.error(f'SSO requested for asset {asset_id} was invoked with invalid method')
            self.logger.info(
                f'Redirecting SSO request for asset {asset_id} to default site',
            )

        return ProductActionResponse.done(http_status=302, headers={'Location': location})
